[0.2] Add additional Linux AF_PACKET options

nathaniel-bennett opened the pull request:
Added socket options, structs and other miscellaneous constants that make memory-mapped packet sockets possible in Linux--PACKET_TX_RING, PACKET_RX_RING, etc.

(adding this to libc 0.2 as 0.3 is in early development)

JohnTitor commented:
Sorry but could you add cfgs here? 0.2's MSRV is 1.13 and we need to add them, see https://github.com/rust-lang/libc/tree/libc-0.2#rust-version-support.

nathaniel-bennett commented:
Added cfgs for union types in latest force-push

JohnTitor commented:
CI fails, seems you also need `libc_align` cfg.

nathaniel-bennett commented:
Added libc_align cfg to any structs that use `repr(align)`

JohnTitor commented on src/unix/linux_like/linux/uclibc/no_align.rs:
```diff
@@ -49,5 +49,5 @@ macro_rules! expand_align {
             }
         }
-    }
+    };
```
Seems this diff is unintended.

nathaniel-bennett commented:
Removed the  unintentional semicolon and rebased

nathaniel-bennett commented:
Rebased to working CI and ready for review

nathaniel-bennett commented:
r? @tgross35

tgross35 commented:
This is a big-ish API surface so it's still taking me a while to going through it. Feel free to hold off rebasing until I get back to you.

tgross35 commented on libc-test/build.rs:
```diff
@@ -3690,6 +3690,23 @@ fn test_linux(target: &str) {
         if (gnu && sparc64) && (ty == "ip_mreqn" || ty == "hwtstamp_config") {
             return true;
         }
+        // FIXME: pass by value for structs that are not an even 32/64 bits on
+        // big-endian systems corrupts the value for unknown reasons.
+        if (sparc64 || ppc || ppc64 || s390x)
+            && (ty == "sockaddr_pkt"
+                || ty == "tpacket_auxdata"
+                || ty == "tpacket_hdr_variant1"
+                || ty == "tpacket_req3"
+                || ty == "tpacket_stats_v3"
+                || ty == "tpacket_req_u")
+        {
+            return true;
+        }
+        // FIXME: musl doesn't compile with `struct fanout_args` for unknown reasons.
+        if musl && ty == "fanout_args" {
+            return true;
+        }
```
What are these unknown reasons, or what is the error? Ideally we link actual rust-lang/rust issues here so we know when this can be removed.

tgross35 commented on src/unix/linux_like/linux/align.rs:
```diff
@@ -200,6 +218,8 @@ macro_rules! expand_align {
                 pub af: u32,
                 pub data: [u8; CANXL_MAX_DLEN],
             }
+
+
```
Looks spurious

tgross35 commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -1314,6 +1442,36 @@ cfg_if! {
             }
         }
 
+        #[cfg(libc_union)]
+        impl ::fmt::Debug for tpacket_req_u {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                f.debug_struct("tpacket_req_u")
+                    .field("req3", unsafe { &self.req3 })
+                    .finish()
+            }
+        }
+
+        #[cfg(libc_union)]
+        impl ::fmt::Debug for tpacket_bd_header_u {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                f.debug_struct("tpacket_bd_header_u")
+                    .field("bh1", unsafe { &self.bh1 })
+                    .finish()
+            }
+        }
+
+        #[cfg(libc_union)]
+        impl ::fmt::Debug for tpacket_block_desc {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                f.debug_struct("tpacket_bd_header_u")
+                    .field("version", &self.version)
+                    .field("offset_to_priv", &self.offset_to_priv)
+                    .field("hdr", &self.hdr)
+                    .finish()
+            }
+        }
+
+
```
Reading `req3` on `tpacket_req_u` reads potentially uninitialized data if the variant is actually `req`. It is probably best to just drop these `Debug` implementations entirely, considering they will likely be removed anyway in 1.0.

nathaniel-bennett commented:
Linked to (https://github.com/rust-lang/rust/issues/43894), removed the spurious whitespace, and dropped the `Debug` implementations. This is a backport of a PR that was already added to the 1.0 branch, so I'll open up a PR taking out the bad `Debug` implementations there too

nathaniel-bennett commented:
@rustbot label S-waiting-on-review

nathaniel-bennett commented:
@rustbot label -S-waiting-on-author
